onnxruntime/test/ir/flatbuffer_utils_test.cc

```diff
@@ -0,0 +1,395 @@
+    } else {
+      ASSERT_EQ(fbs_tensor->external_data_offset(), -1) << "external_data_offset is set for string data when we expect it to not be set for tensor " << fbs_tensor->name()->str();
+      ASSERT_TRUE(fbs_tensor->raw_data() || fbs_tensor->string_data()) << "tensor has no data attached to it" << fbs_tensor->name()->str();
+    }
+  }
+
+  bool data_validated = true;
+
+  // initializers = expected, in the form of tensorproto
+  // loaded_initializers = actual, in the form of tensorproto
+  ASSERT_EQ(initializers.size(), loaded_initializers.size());
+
+  for (int i = 0; i < static_cast<int>(initializers.size()); i++) {
+    const auto& expected_initializer = initializers[i];
+    const auto& loaded_initializer = loaded_initializers[i];
+    // validate the loaded initializer
+    ASSERT_EQ(expected_initializer.name(), loaded_initializer.name());
+    ASSERT_EQ(expected_initializer.data_type(), loaded_initializer.data_type());
+    ASSERT_EQ_TENSORPROTO_VECTORFIELD(expected_initializer, loaded_initializer, dims());
+    if (loaded_initializer.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      // extract expected tensor raw data
+      std::vector<uint8_t> expected_data;
+      ASSERT_STATUS_OK(onnxruntime::utils::UnpackInitializerData(expected_initializer, expected_data));
+
+      ASSERT_EQ(expected_data.size(), loaded_initializer.raw_data().size()) << "expected initializer name " << expected_initializer.name() << " | loaded initializer name " << loaded_initializer.name();
+      std::vector<uint8_t> loaded_data(loaded_initializer.raw_data().begin(), loaded_initializer.raw_data().end());
+      for (int j = 0; j < static_cast<int>(expected_data.size()); ++j) {
+        ASSERT_EQ(expected_data[j], loaded_data[j]) << "expected initializer name " << expected_initializer.name() << " | loaded initializer name " << loaded_initializer.name();
+      }
+    } else {
+      // string type tensor
+      ASSERT_EQ_TENSORPROTO_VECTORFIELD(expected_initializer, loaded_initializer, string_data());
+    }
+  }
+  ASSERT_TRUE(data_validated);
+}
+
+#ifdef ENABLE_TRAINING_APIS
+// tests method that loads to OrtTensor (used when loading a checkpoint into a checkpoint state)
+TEST(GraphUtilsTest, ExternalWriteReadWithLoadOrtTensor) {
+  // create data
+  auto initializers = CreateInitializersNoString();
+
+  flatbuffers::FlatBufferBuilder builder(1024);
+
+  // write
+  std::ofstream output_stream;
+  ExternalDataWriter writer;
+  CreateWriter("ExternalWriteReadBasicTest.bin", output_stream, writer);
+
+  std::vector<flatbuffers::Offset<fbs::Tensor>> fbs_tensors;
+  for (const auto& initializer : initializers) {
+    flatbuffers::Offset<fbs::Tensor> fbs_tensor;
+    ASSERT_STATUS_OK(SaveInitializerOrtFormat(builder, initializer, Path(), fbs_tensor, writer));
+    fbs_tensors.push_back(fbs_tensor);
+  }
+
+  ASSERT_EQ(output_stream.tellp(), 840) << "Data written to the external file is incorrect.";
```

Comment:
Try to avoid magic numbers in tests. It requires more maintenance work to update the test, and it can also be unclear how the original number was proven to be correct.

In this case it's probably not possible though due to there being both padding for alignment and the unpack of the data potentially expanding it.